### DISCUSSION TITLE
WASM NPM module

### DISCUSSION THREAD

**Discussion by tristan-f-r** [pest-parser/pest#773] [Ideas]
Having official support for an NPM/browser/deno module (ex [lightningcss](https://github.com/parcel-bundler/lightningcss/tree/master/wasm) does this) would be awesome.

**Comment by tomtau** [comment]
It should be relatively straightforward to do with wasm-pack: https://rustwasm.github.io/docs/wasm-pack/ (as that's already done for the website: https://github.com/pest-parser/site )

The main question is what the potential package bindings should contain. Do you have any thoughts about it?

"pest_vm" may be a good candidate to expose (that's what the website uses), but I don't think it is that widely useful. Most external projects use the compile-time approach ("pest" + "pest_derive"). In that approach, I guess the most natural approach in the Wasm context is that each external
project can create their own bindings to their own specific language (they implemented in Rust and pest), i.e. there is no need for a NPM package?

**Reply by tristan-f-r** [reply]
Pest's vm has good usage inside other potential ventures -- for example, if pest ever [supports a VSCode extension](https://github.com/pest-parser/pest/discussions/771), in order to support the web, there needs to be a WASM-compatible version of pest that can run within it. I also do agree with the latter, but exposing pest_vm would be useful for any future web meta-pest projects.

**Reply by tomtau** [reply]
@LeoDog896 VSCode extension has these language server bindings: https://github.com/pest-parser/pest-ide-tools/blob/main/language-server/src/lib.rs do you think that's worth publishing as an NPM package, or it's better to have more "vanilla" pest_vm bindings (similar to what could be used on the website)?

**Reply by tristan-f-r** [reply]
We use a diff. version in `site` too, right? It may be better to publish a standalone NPM package.

**Reply by tristan-f-r** [reply]
Also, I would like to mention that having pest as an NPM package would be very useful for integration into editors - e.g. CodeMirror or monaco-editor.
